### ISSUE TITLE
Cluster & Edit results sort impovement

### ISSUE THREAD

**Issue by belg4mit** [OpenRefine/OpenRefine#2101]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**
The Cluster & Edit results are sorted by decreasing number of existing terms in a cluster, which is almost entirely useless default.

**Describe the solution you'd like**
Sorting by decreasing sum of affected rows would be a more useful behavior, as it means one could more easily focus attention on verifying clusters that will have the largest impact on the data.

**Describe alternatives you've considered**
Ideally, one should be able to click on any of the headers to sort by that column.

**Additional context**
N/A

**Comment by wetneb** [comment]
Note that it is currently possible for clustering algorithms to return clusters with only one value (possibly present on multiple rows). I am not sure why this choice was made - surely there is nothing to merge in these cases? Anyway, if the change proposed above is implemented, it would probably be necessary to filter out these spurious clusters (that might be a good thing to do regardless of this issue I would say).

**Comment by wetneb** [comment]
The odd behaviour I noticed above should be fixed, see #2152. I still agree with the issue here: we should change the ordering of the results.

**Comment by jamessspanggg** [comment]
@wetneb I want to work on this but I noticed that for kNNClusterer, no sorting is being done. (as compared to BinningClusterer where it sorts at the end). Should I add a sort to the kNNClusterer as well?

https://github.com/OpenRefine/OpenRefine/blob/3cb7525d49d3e7029e45940988221effb9f825fe/modules/core/src/main/java/com/google/refine/clustering/knn/kNNClusterer.java#L238-L245

https://github.com/OpenRefine/OpenRefine/blob/3cb7525d49d3e7029e45940988221effb9f825fe/modules/core/src/main/java/com/google/refine/clustering/binning/BinningClusterer.java#L229-L237

**Comment by wetneb** [comment]
@jamessspanggg oh interesting! Is there no sorting done downstream either? If so then that sounds really bad. I guess we could ideally do this sorting only in one place, probably independently from the clusterer being used, if the criterion is always the same?

Sorting by decreasing number of affected rows would be ideal both for KNN and Fingerprint clusterers, I would say.

**Comment by jamessspanggg** [comment]
> @jamessspanggg oh interesting! Is there no sorting done downstream either? If so then that sounds really bad. I guess we could ideally do this sorting only in one place, probably independently from the clusterer being used, if the criterion is always the same?

Yes the criterion is the same, however the structure of the two clusters are different. 

In BinningClusterer, an `ArrayList<Map<String, Integer>>` is used for clusters that maps each row name to its number of affected rows. 

https://github.com/OpenRefine/OpenRefine/blob/5cf332c981bfc88e185eacb15c88906dfb3d57ba/main/src/com/google/refine/clustering/binning/BinningClusterer.java#L128

In KNNClusterer, a `List<Set<Serializable>>` is used for clusters and a separate `Map<Serializable, Integer>` is used to store the number of affected rows for each row name.

https://github.com/OpenRefine/OpenRefine/blob/5cf332c981bfc88e185eacb15c88906dfb3d57ba/main/src/com/google/refine/clustering/knn/kNNClusterer.java#L128-L130

In order to sort the clusters in KNNClusterer, the data structure of the cluster might need to be changed to include the corresponding number of affected rows.

**Comment by wetneb** [comment]
I think it would make sense to unify this, but it's completely up to you. Targeted fixes are welcome too!

If I had to redesign this, I think the `Clusterer` interface could be improved. At the moment clusterers are just expected to be serializable in JSON via Jackson and that is how the computed clusters are exposed. I would say there should be something more explicit in the `Clusterer` interface to retrieve the clustering results - either by changing the return type of `computeClusters` to return the clusters directly, or by providing another method to get the clustering results after `computeClusters` returned. That might require or benefit introducing dedicated class(es) for the clustering results instead of nesting maps and lists.

But again if that feels too involved feel free to go for a small fix instead.

**Comment by tfmorris** [comment]
@jamessspanggg Are you still planning to work on this?

**Comment by tfmorris** [comment]
> I noticed that for kNNClusterer, no sorting is being done. (as compared to BinningClusterer where it sorts at the end)

The results are returned from Vicino already sorted, so no sorting is required.

> I would say there should be something more explicit in the Clusterer interface to retrieve the clustering results - either by changing the return type of computeClusters to return the clusters directly, or by providing another method to get the clustering results after computeClusters returned. 

As was pointed out, the two different clusterers return results in slightly different formats, which I suspect is why the getClusters method (which exists) wasn't added to the interface. That's something we can probably sort out when we do work in this space.
